crypto/ec/ec_lib.c

```diff
@@ -852,7 +853,10 @@ int EC_POINT_is_on_curve(const EC_GROUP *group, const EC_POINT *point,
         ECerr(EC_F_EC_POINT_IS_ON_CURVE, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
         return 0;
     }
-    if (group->meth != point->meth) {
+    if (group->meth != point->meth
+            || (group->curve_name != 0
+                && point->curve_name != 0
+                && group->curve_name != point->curve_name)) {
         ECerr(EC_F_EC_POINT_IS_ON_CURVE, EC_R_INCOMPATIBLE_OBJECTS);
```

Comment:
While this short circuit check doesn't seem to harm, is it really of practical use? For example, if the point is received from a peer during a Diffie-Hellman key exchange, then the full check is necessary anyway.

Reply:
The problem in #6302 came about because we made assumptions about the internal encoding of a point when checking if a point was on the curve (IIUC we were checking if a point whose co-ordinates were encoded relative to one montgomery context were on the curve whose points would be expected to be encoded relative to a different montgomery context). Now we fixed that issue, but it seems to me that it would be easy to make that same mistake in lots of other places. This was an attempt to sidestep the whole problem and not start trying to do anything with a point if it is for a different curve.

Reply:
Ah, I see. So in the case of a DH keyexchange the (x,y) coordinates from the peer are passed to EC_POINT_set_affine_coordinates_GFp() before reaching this check, hence point->curve_name is already set to the expected curve type, right?

Reply:
Right - when we first create an "empty" EC_POINT object we have to supply the EC_GROUP it is associated with. Assuming the EC_GROUP isn't a custom curve it will have a curve name associated with it, so we save that away. When we later set the co-ords for it we know that those co-ords will be encoded internally in a manner appropriate to the group. If we later did a check for it being on a curve we can immediately sanity check that we're on the right curve before we start  - so we won't even try and do operations on points encoded in a different way to what we expect.

This only works for named curves. But I think this is the vast majority of cases anyway. I also take @davidben's point that actually this is only one place where we check for consistency. If we're going to do this, then we should do it everywhere that a consistency check is currently (or not at all).